rfcs/2020-07-21-2744-vicscript.md

```diff
@@ -0,0 +1,191 @@
+# RFC #2744 - Vicscript
+
+## Motivation
+
+Simple transforms are easy to document, understand and recommend, powerful transforms are more likely to solve a use case in a concise way. We currently have two tiers of transform at opposite ends of this spectrum:
+
+- Pure TOML (`add_fields`, `remove_fields`, `json_parser`)
+- Turing complete runtime (`lua`, the new WASM transform)
+
+This gives us decent coverage for very simple use cases where only one or two native transforms are needed, and also with the super advanced cases where a user basically just needs a framework for running a program they intend to write and maintain themselves.
+
+However, there's a common middle ground where more complex and oftentimes conditional mappings are needed. This is where our native transforms become cumbersome to work with, and yet a full runtime would be heavy handed and is difficult to provide support for (or often simply doesn't perform well enough).
+
+The main motivation with Vicscript (temporary name) is to introduce a tool that strikes a better balance by offering a simple language that isn't turing complete and runs as fast as our native transforms. Something that's easy to document, understand, use, diagnose and fix.
+
+My proposal here is basically just a copy of the bits I think work well from [Bloblang](https://www.benthos.dev/docs/guides/bloblang/about), which itself is a derivative of [IDML](https://idml.io/). This is partly due to my familiarity with these languages, but I also have first hand experience of seeing them in action and know that they are easy to implement and adopt.
+
+### Out of Scope
+
+Event expansion and reduction are within the scope, but routing events is not. For example, it would be possible to express with vicscript that an event should be expanded into several, and then at the sink level filter events (potentially using vicscript again) so that they route to different destinations based on their contents. However, vicscript itself will not have any awareness of sinks, only events.
+
+Joining events are within scope, but aggregation of those events is not. For example, it would be possible to aggregate events using a transaction transform (or wasm, etc) in a form where vicscript can be used to express exactly how the fields of those events should be joined. However, vicscript itself will not have the capability to temporarily store events and has no awareness of delivery guarantees or transactions.
+
+## Guide Level Proposal
+
+The `vicscript` transform allows you to map your events using Vicscript. A Vicscript statement has a left hand side target path and a right hand side query:
+
+```toml
+[transforms.mapper]
+  inputs = [ "foo" ]
+  type = "vicscript"
+  mapping = """
+foo = "hello"
```

Comment:
Since this is the first presentation of this syntax, the thing that immediately jumps out to me is the lack of clarity around _event_ field assignment. It looks like a bunch of variables are being set with no real action. Following precedence set by our `lua` transform, I'm curious what you think about this alternative syntax?

```
event.log.foo = "hello"
event.log.bar = event.log.bar + 10
event.log.baz_is_large = event.log.baz_is_large > 10
```

I realize the intent of your syntax is to avoid maintaining a runtime environment. So things like this would _not_ be possible:

```
# not possible
baz = 1
event.log.bar = baz
```

And we could simply throw an error saying that `baz` is "unknown". The advantage with my syntax is:

1. It's obvious you are setting an event level field.
2. The `log` namespace means we can use this with metrics (unless you were thinking of a better way to solve this).
3. It's closer in syntax to other transforms that manipulate events (like `lua` and `wasm`).

Reply:
I think that'd be reasonable, one thing to note is that enforcing a path prefix would be stripped within a map definition, so you'd still get bare assignments like this:

```coffee
map foo {
  type = (type.uppercase() | "UNKNOWN")
  name = name.titlecase()
}

event.log.metadata = event.log.metadata.apply("foo")
```

Another thing to note is that both IDML and Bloblang have an optional path prefix `root` (`root.foo = "bar"` would set the event field `foo` to `"bar"`), which is useful for when a path clashes with a keyword like `if` or `match`, but is also implicit when omitted as a convenience. It's assumed it usually won't be used unless you specifically want to assign directly to the root. On the right hand side we have a similar keyword `this` which represents the source value in the current context.

Adding them to the above example looks like this:

```coffee
map foo {
  root.type = (this.type.uppercase() | "UNKNOWN")
  root.name = this.name.titlecase()
}

root.event.log.metadata = this.event.log.metadata.apply("foo")
```

I would opt to leave these in the spec as they are. We could choose to make them mandatory but I wouldn't rename `root` to something like `event` as it's somewhat misleading in the context of maps.

On the topic of variables, we can actually support them without a runtime or making the spec turing complete, it's just a mapping that isn't part of the output event. I originally omitted them from the RFC so we could consider them later as a more advanced feature, Bloblang has them and it's definitely useful, I already have an example [here](https://github.com/timberio/vector/blob/vicscript-rfc/rfcs/2020-07-21-2744-vicscript/example1.coffee) that uses them.

Reply:
> I would opt to leave these in the spec as they are.

Sounds good. I've come around to leaving out the prefix, especially as we start to incorporate it in other places where specifying `event.log` would be confusing. Such as dynamic configuration values.

> ```coffee
> map foo {
>   type = (type.uppercase() | "UNKNOWN")
>  name = name.titlecase()
> }
> ```

The `map` syntax feels a little strange to me as well. As I noted in my other comment, we should agree on these details in the form of a spec so that we can obtain consensus. I realize that's the part of what this RFC aims to accomplish, but it still feels hand-wavy to me.

> On the topic of variables, we can actually support them without a runtime or making the spec turing complete, it's just a mapping that isn't part of the output event.

Sounds good to me.

Reply:
Yeah we can do that. One thing that's open to confusion here is that `.` on the left hand side means something different from the right hand side.

Using `root` and `this` makes that distinction clear (even if it's not immediately obvious what they point to), whereas a very common pitfall of jq is that users will naturally assume in many places that a dot on the left and the right are the same thing. It means the docs for lots of features ends up needing to also explain that same concept continuously, making them quite verbose: https://stedolan.github.io/jq/manual/#Assignment.

Reply:
> One thing that's open to confusion here is that `.` on the left hand side means something different from the right hand side.

Yeah, this is interesting. It looks like in "normal" `=` assignment they are the same thing, but "most users will want to use modification assignment operators", where the RHS is a filter. I get the feeling there's some history there and I'm curious what led to the current state.